### ISSUE TITLE
Compiler Bug when running tests

### ISSUE THREAD

**Issue by rajkumarGosavi** [vlang/v#18121]
### Describe the bug

When executing the test cases I get some random error in my terminal which does not describe the cause.

### Expected Behavior

Since it says that its a compiler error a clean concise error description that is actually helpful is expected.

### Current Behavior

irrelevant details in terminal

```
---- Testing... -----------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------
==================
/tmp/v_1000/tsession_7fbddbf08740_12036335798954/repo_test.16320955974902252362.tmp.c:16258: error: '_t3' undeclared
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang


---------------------------------------------------------------------------------------------------------------------
Failed command 1:    v' -stats  -o '/tmp/v_1000/tsession_7fbddbf08740_12036335798954/repo_test' 'src/repo/repo_test.v'
Summary for all V _test.v files: 1 failed, 1 total. Runtime: 619 ms, on 1 job.
```

### Reproduction Steps

clone https://github.com/rajkumarGosavi/todo-app

do `v -stats test . `

### V version

Current V version: V 0.3.4 134e781

### Environment details (OS name and version, etc.)

OS: linux, Linux Mint 21.1

**Comment by Casper64** [comment]
Is `TodoStore` an interface? If it's a struct I think you forgot to use { brackets instead of ( brackets when creating `ts`.

**Comment by Casper64** [comment]
I see that you are using `TodoStore` as an type alias to an array. I don't think you can add methods to a type like that. 

If you convert `TodoStore` to a struct and change the methods to use `ts.todos` it should work.

```v
struct TodoStore {
pub mut:
	todos []todo.Todo
}
```

**Comment by rajkumarGosavi** [comment]
After following your advice, here is the branch  https://github.com/rajkumarGosavi/todo-app/tree/test-fix
with the necessary changes.

I am still getting the same error

```
---- Testing... -----------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------
==================
/tmp/v_1000/tsession_7f9f93444740_20666925572502/repo_test.15305419493342423413.tmp.c:16264: error: '_t3' undeclared
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

**Comment by rajkumarGosavi** [comment]
Also @Casper64 for 
> I see that you are using TodoStore as an type alias to an array. I don't think you can add methods to a type like that.

Since i am creating a type alias which is my own type I should have full control over the type.

In Go we can do that, do we have a work around for it?

**Comment by JalonSolov** [comment]
You are correct - if you create an alias, you are supposed to have full control over the alias, adding methods, etc.

**Comment by Casper64** [comment]
The first bug is in your code. In `repo.v` you check verify `id` backwards

```v
fn (ts TodosStore) get_todo(id int) ?todo.Todo {
	// if id is 5 and len is 1 `none` is not returned!
	if ts.len >= id {
		return none
	}

	return ts[id - 1]
}
```

change it to 
```v
fn (ts TodosStore) get_todo(id int) ?todo.Todo {
	if id >= ts.len {
		return none
	}

	return ts[id - 1]
}
```

But there is an actual bug: it's in `test_get_todo`.

```v
assert ts.get_todo(5) == none
```

For some reason you can't directly check if a value is `none` with assert.  But if you store the variable first and then assert the test succeeds.

```v
val := ts.get_todo(5)
assert val == none
```

If I look at the C code it seems that the assert option is validated before the assert check is executed and that results in `_t3` being undefined.

```c
	_option_src__todo__Todo _t4;
	// _t3 is defined on the next line
	_option_ok(&(src__todo__Todo[]) { _t3 }, (_option*)(&_t4), sizeof(src__todo__Todo));
	 _t4;
	// assert
	_option_src__todo__Todo _t3 = src__repo__TodosStore_get_todo(ts, 5);
```

**Comment by Casper64** [comment]
This issue is related to #16410 and #10803 which are still not resolved, but you can see the comments of alternetive ways to check for an error.

**Comment by rajkumarGosavi** [comment]
Thank you for pointing out the length check problem.

Moving the value to a variable first and then asserting solves the issue. 

Some points that can help someone learning V or coming from a different language 
1. Better error message
2. There were no errors shown by the `vls` in the test case, if the `vls` would have identified that direct assert is not possible on Option then the developer could have directly fixed it at that place itself. I believe this will be resolved in future plan.
3. Not sure that the team should explicitly mention it in the docs, `that you cannot directly use the Option and we have to basically unwrap it to a variable first`.


I hope the bug that was found is in the compiler, and not the user code.

Please feel free to close this issue if required. 

It was very helpful communicating with you @Casper64. Appreciate the quick resolution.

Regards!!
